### PR TITLE
Crc us smooth crtmask pipeline

### PR DESCRIPTION
**Pull request by ChristophePhillips**
This includes a few important fixes regarding the spatial processing. More specifically:
- keeps the smoothed (modulated warped) tissue maps ▶️ to be used in a a standard VBM
- adds a function for the creation of the masks used in the statistical analysis. This is further included in the pipeline
- in the tissues-weighted smoothing function, ensures that masked out area gets a NaN value if images are in a float format + adds a small ad hoc function that can do this a posteriori.

All these are included in the MatlabBatch and config files.

### CODE REVIEW

File: hmri_check_opt.m [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+%
+% Function to automatically check the content of a "opts" structure, using
+% a "default opts structure", adding the missing fields and putting in the 
+% default value if none was provided.
+%
+% INPUT:
+% opts_def  : default or reference structure
+% opts      : input option structure that need to be filled for missing
+%             fields with default values
+%
```
Comment (lukeje) [inline hmri_check_opt.m:13]: `loc_opt` input not documented.

File: hmri_proc_MPMsmooth.m [hunk 2]
```diff
@@ -84,16 +89,22 @@
 end
 
 % Flags for image calculation
-ic_flag = struct(...
-    'dtype', 16, ... % keep things in floats
+ic_flag = struct(...%    'dtype', 16, ... % keep things in floats
```
Comment (lukeje) [inline hmri_proc_MPMsmooth.m:92]: What does this do? Does it just maintain the input data type?

File: hmri_proc_MPMsmooth.m [hunk 3]
```diff
@@ -105,12 +116,14 @@
...
     
-    % Smooth TC -> m-images
+    % Smooth TC -> m-images, generate at 1st run
+    if ii==1
     m = cell(nTC,1);
     for jj=1:nTC
         m{jj} = spm_file(fn_mwTC(jj,:),'prefix','s','number','');
         spm_smooth(fn_mwTC(jj,:),m{jj},fwhm); % smooth mwc(jj)
     end
+    end
```
Comment (lukeje) [inline hmri_proc_MPMsmooth.m:126]:
```suggestion
    if ii==1
        m = cell(nTC,1);
        for jj=1:nTC
            m{jj} = spm_file(fn_mwTC(jj,:),'prefix','s','number','');
            spm_smooth(fn_mwTC(jj,:),m{jj},fwhm); % smooth mwc(jj)
        end
    end
```

File: hmri_proc_crtMask.m [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+% 
+% FORMAT
+%   [fn_maskTC,fn_meanTC] = hmri_proc_crtMask(fn_smwTC, opts)
+% 
+% INPUT
+% - fn_smwTC : filenames (cell array, 1 cell per TC, containing a char or 
+%             cell array) of the modulated warped tissue classes, 
+%             ex. the mwc1/2*.nii filenames in the 1st and 2nd cell.
+% - opts    : option structure
+%   .minTCp : minimal posterior probability value to be considerd [.2, def]
```
Comment (lukeje) [inline hmri_proc_crtMask.m:15]:
```suggestion
%   .minTCp : minimal posterior probability value to be considered [.2, def]
```

File: hmri_proc_crtMask.m [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+%   .noOvl  : avoid overlap by assigning to tissue class with maximum
+%             posterior probability [1, def]
+%   .outPth : path to output folder, use that of the 1st image if it is 
+%             left empty ['', def]
+% 
+% OUTPUT
+% - fn_maskTC : filenames (char array) of the tissue specific masks
+% - fn_meanTC : filenames (char array) of the mean tissue class images
+% 
+% The core idea comes from Callaghan et al.
```
Comment (lukeje) [inline hmri_proc_crtMask.m:25]: Year, journal for Callaghan, et al.?

File: hmri_proc_crtMask.m [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+% > smoothed Jacobian-modulated tissue probability maps in MNI space were  
+% > averaged across all subjects for each tissue class (GM, WM, and 
+% > cerebrospinal fluid). Masks were generated by assigning voxels to the 
+% > tissue class for which the probability was maximal. Voxels for which 
+% > neither the GM nor the WM probability exceeded 20% were excluded from 
+% > the analysis. This approach was used to ensure that each voxel was 
+% > analyzed in only one subspace and that non-brain tissue was excluded.
+% 
+% The function is relatively flexible as
+% - nTC differe tissue class images can be introduced. if only one is
```
Comment (lukeje) [inline hmri_proc_crtMask.m:35]:
```suggestion
% - nTC different tissue class images can be introduced. If only one is
```

File: hmri_run_proc_pipeline.m [hunk 7]
```diff
@@ -143,14 +156,39 @@
 % Run the *_proc_smooth
 fprintf('\nhMRI-pipeline: running the weighted-average (smoothing) module.\n')
 out_wa = hmri_run_proc_smooth(job_smooth);
-% where the 'out_wa' is organized as a structure out_wa.tc where
-% - tc is a cell-array of size {n_TCs x n_pams}
-% - each element tc{ii,jj} is a cell array {n_subj x 1} with each subject's
-%   smoothed data for the ii^th TC and jj^th MPM
+% The 'out_wa' structure is organized as a structure with 2 fileds
```
Comment (lukeje) [inline hmri_run_proc_pipeline.m:159]:
```suggestion
% The 'out_wa' structure is organized as a structure with 2 fields
```

File: hmri_run_proc_smooth.m [hunk 8]
```diff
@@ -4,10 +4,13 @@
 % Data are selected in a 'many subject' style, i.e. all the images of one
 % type are selected from many subjects at once!
 % 
-% The 'out' structure is organized as a structure out.tc where
-% - tc is a cell-array of size {n_TCs x n_pams}
-% - each element tc{ii,jj} is a cell array {n_subj x 1} with each subject's
-%   smoothed data for the ii^th TC and jj^th MPM
+% The 'out' structure is organized as a structure with 2 fileds
```
Comment (lukeje) [inline hmri_run_proc_smooth.m:7]:
```suggestion
% The 'out' structure is organized as a structure with 2 fields
```

File: tbx_scfg_hmri_proc.m [hunk 9]
```diff
@@ -33,14 +33,16 @@
 proc_modul.help    = {
     ['Parameter maps are spatially processed and brought into standard space',...
     'for furhter statistical analysis.']
```
Comment (lukeje) [inline tbx_scfg_hmri_proc.m:35]:
```suggestion
    'for further statistical analysis.']
```

File: tbx_scfg_hmri_proc.m [hunk 10]
```diff
@@ -33,14 +33,16 @@
...
-    ['This include 3 main processing steps:']
+    ['This include 4 main processing steps:']
```
Comment (lukeje) [inline tbx_scfg_hmri_proc.m:36]:
```suggestion
    ['This includes 4 main processing steps:']
```

File: tbx_scfg_hmri_proc.m [hunk 11]
```diff
@@ -50,18 +52,19 @@
 proc.name    = 'Process hMRI maps';
 proc.help    = {
     ['Parameter maps are spatially processed and brought into standard space',...
-    'for furhter statistical analysis.']
-    ['This include 3 main processing steps:']
+    'for further statistical analysis.']
+    ['This include 4 main processing steps:']
```
Comment (lukeje) [inline tbx_scfg_hmri_proc.m:56]:
```suggestion
    ['This includes 4 main processing steps:']
```

File: tbx_scfg_hmri_proc_pipeline.m [hunk 12]
```diff
@@ -166,13 +169,17 @@
...
     [' ']
     ['For simplicity, 2 standard pipelines are also set up:']
-    ['- US+Smooth -> applies US, warps into MNI, then smoothes (weighted-average)']
-    ['- US+Dartel+Smooth -> applies US, builds Dartel template and warps' ...
-    'into MNI, then smoothes (weighted-average)']
+    ['- US+Smooth+MaskCrt -> applies US, warps into MNI, smoothes (weighted-average), then creates tissue masks']
+    ['- US+Dartel+Smooth+MaskCrt -> applies US, builds Dartel template and warps ' ...
+    'into MNI, smoothes (weighted-average), then creates tissue masks']
+    ['Along with tissue specific masks, the mean smoothed tissue class images are also created. ' ...
+    'The masks should then be enterd as ''explicit mask'' ' ...
```
Comment (lukeje) [inline tbx_scfg_hmri_proc_pipeline.m:180]:
```suggestion
    'The masks should then be entered as ''explicit mask'' ' ...
```

**Review by lukeje — COMMENTED** [review]
Hey Chris,

I made some stylistic suggestions (typos in the comments and indentation suggestions), but it looks reasonable to me. I didn't test it because I'm not sure whether I have the right kind of data. I'll ask at the pull request meeting whether anyone there has tested your changes yet.

File: hmri_proc_MPMsmooth.m (hunk 2 above)
Comment (ChristophePhillips) [reply hmri_proc_MPMsmooth.m:92]: Since most (all?) the maps used are encoded in a float format, setting the `dtype` flag to 16 forces `spm_imcalc` to write out the resulting image in float format as well. Otherwise the default format will be in `int16`, see [here](https://github.com/spm/spm12/blob/master/spm_imcalc.m#L21). 

A more elegant way would be to check the format of the input images and to adjust the output format accordingly... Shall we add an issue about this?

File: hmri_check_opt.m (hunk 1 above)
Comment (ChristophePhillips) [reply hmri_check_opt.m:13]: It is described actually but, admittedly not at the right place, see around lines 22-23.

File: hmri_proc_MPMsmooth.m (hunk 2 above)
Comment (lukeje) [reply hmri_proc_MPMsmooth.m:92]: OK, so the commented out part in the new version means that `int16` will be used? Is this what you intended?

File: hmri_proc_crtMask.m (hunk 5 above)
Comment (ChristophePhillips) [reply hmri_proc_crtMask.m:25]: 2014 https://doi.org/10.1016/j.neurobiolaging.2014.02.008

File: hmri_proc_MPMsmooth.m (hunk 2 above)
Comment (ChristophePhillips) [inline hmri_proc_MPMsmooth.m:92]:
```suggestion
ic_flag = struct(... % type is set below based on that of input image
```

File: hmri_proc_crtMask.m [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+%   .noOvl  : avoid overlap by assigning to tissue class with maximum
+%             posterior probability [1, def]
+%   .outPth : path to output folder, use that of the 1st image if it is 
+%             left empty ['', def]
+% 
+% OUTPUT
+% - fn_maskTC : filenames (char array) of the tissue specific masks
+% - fn_meanTC : filenames (char array) of the mean tissue class images
+% 
+% The core idea comes from Callaghan et al.
```
Comment (ChristophePhillips) [inline hmri_proc_crtMask.m:25]:
```suggestion
% The core idea comes from Callaghan et al. 2014
% https://doi.org/10.1016/j.neurobiolaging.2014.02.008
```

**Comment by micahgallen** [comment]
Hello!

Sorry for my github naivety, but is it correct that the explicit mask creation function has not been merged to the toolbox? We are thinking to use it for the explicit mask creation in our VBQ multiple regression analysis. For now it looks like we could just directly use the proc_crtMask.m function on our wap1 etc images? Then apply the derived maps as explicit maps (with no implicit or threshold masking) in the model creation? 

Sorry if this is the wrong place... can cross post to the mailing list if needed. 

Cheers,
Micah

**Comment by ChristophePhillips** [comment]
Hi @micahgallen 

Yes, indeed it is NOT fully integrated yet... and I am pleading guilty on this one.
I had a preliminary solution implemented but then before merging we reviewed the code and found a few elements that could have been more carefully implemented. Plus we are trying to make sure new bits of code are automatically tested.
All in all, the end of academic year was particularly busy and I am (long) behind schedule on these tasks. Hopefully dealt with by end of September or October.

TL/DR: no it is not implemented as it needs improvements, still current code available on in this branch should work correctly.

**Comment by micahgallen** [comment]
Thanks so much for your speedy response, and for all the hard work on this great toolbox!

**Comment by ChristophePhillips** [comment]
@lukeje & @halirutan 
I just added the test function for the zero-to-nan conversion, used to ensure the correct implicit masking with float-formatted data. Hopefully we can conclude this PR at next Hackathon meeting. 😃
